board: m5stack-core2 module support

MrMarteng opened the pull request:
This Pull-Request features support for the [M5Stack Core2](https://docs.m5stack.com/en/core/core2) hardware. The module features a LCD display, touchscreen, battery, USB-Port, 8MB RAM and 16MB Flash.
Basic features are working. See board documentation for details.

Signed-off-by: Martin Kiepfer <mrmarteng@teleschirm.org>

Commit:
boards: M5Stack Core2 module suppport

This Pull-Request enables support for the M5Stack Core2 hardware.
The module features a LCD display, touchscreen, battery, USB-Port,
8MB RAM and 16MB Flash.
Basic features are working. Please refere to the board documentation
for details.

Signed-off-by: Martin Kiepfer <mrmarteng@teleschirm.org>

gpio: gpio_axp192: fix dependecy between GET_DIRECTION and GET_CONFIG
configuration

Latest unit gpio get_direction unit test failes, as get_direction api
internally requires get_config functionality. This commit fixes this
dependecy.

Signed-off-by: Martin Kiepfer <m.kiepfer@teleschirm.org>

kartben commented:
Very cool! 
It doesn't have to be through this PR but it would be nice to have a binding for the MBUS header, as this will make it easier to start adding M5 shields that "just work" :)

MrMarteng commented:
> MBUS

Can you describe this idea on more detail? Are you referring to the Module BUS M5 is offering (https://docs.m5stack.com/en/module/bus)?

There are generally a lot of other functionalities missing yet (gauge, audio, ...). But I want to add additional functionality in future.

kartben commented:
I mean describing the pin headers similar to how it's done for say Arduino R3 headers, grove connectors, or Raspberry Pi 40-pin headers. You should be able to find examples easily by searching the code base. Thanks to this, shield definitions for M5Stack modules would conveniently be able to reference pins from the header, with Devicetree then automagically "routing" them to the corresponding pins of the ESP32. Plus you could also define aliases like m5stack_i2c or m5stack_spi.

Here's the description of the pin mappings for the M5Stack Core2: https://static-cdn.m5stack.com/resource/docs/products/core/core2/core2_mbus_01.webp

Hope this helps clarify!

MrMarteng commented:
> I mean describing the pin headers similar to how it's done for say Arduino R3 headers, grove connectors, or Raspberry Pi 40-pin headers. You should be able to find examples easily by searching the code base. Thanks to this, shield definitions for M5Stack modules would conveniently be able to reference pins from the header, with Devicetree then automagically "routing" them to the corresponding pins of the ESP32. Plus you could also define aliases like m5stack_i2c or m5stack_spi.
> 
> Here's the description of the pin mappings for the M5Stack Core2: https://static-cdn.m5stack.com/resource/docs/products/core/core2/core2_mbus_01.webp
> 
> Hope this helps clarify!

Thank you. I've understood and like the idea. I will add this functionality in a separate PR.

sylvioalves commented on boards/xtensa/m5stack_core2/Kconfig.defconfig:
```diff
@@ -0,0 +1,55 @@
+	default BT_ESP32 if BT
+endchoice
+
+config I2C
+	default y
```
You could remove this as it is already defined in DTS level as "okay". @kartben, removing this from here would be the recommended way?

MrMarteng replied:
Fixed

sylvioalves commented on boards/xtensa/m5stack_core2/m5stack_core2.dts:
```diff
@@ -0,0 +1,241 @@
+	status = "okay";
+};
+
+&gpio0 {
```
Perhaps you can join both `gpio0` nodes?

MrMarteng replied:
Thank you. Fixed with latest commit.

sylvioalves commented on boards/xtensa/m5stack_core2/Kconfig.board:
```diff
@@ -0,0 +1,8 @@
+
+# Copyright (c) 2023 Martin Kiepfer <m.kiepfer@teleschirm.org>
+# SPDX-License-Identifier: Apache-2.0
+
```
Can you check whether it is missing `SOC_PART_NUMBER`? This would define PSRAM automatically:

```c
choice SOC_PART_NUMBER
	default SOC_ESP32_WROVER_E_N16R8
endchoice
```

MrMarteng replied:
M5Stack-Core2 is featuring SOC_ESP32_D0WD_V3. This is very similar to SOC_ESP32_WROVER_E_N16R8, but I find it more consistent to keep sram and flash definition separately within board layer.

Commit:
boards: m5stack_core2: Exclude board from device tree test

m5stack_core2 enables i2c by default for regulator usage which
conflicts with the test and is therefore excluded from the test.

Signed-off-by: Martin Kiepfer <mrmarteng@teleschirm.org>

MrMarteng commented on tests/lib/devicetree/devices/testcase.yaml:
```diff
@@ -14,4 +14,5 @@ tests:
       - bt610
       - bl5340_dvk_cpuapp
       - bl5340_dvk_cpuapp_ns
+      - m5stack_core2
```
Added a separate commit for this change.

kartben replied:
note was added

MrMarteng commented:
Typo in commit message has been corrected.

MrMarteng commented:
Any further comments, change requests?

marekmatej commented on boards/xtensa/m5stack_core2/m5stack_core2.dts:
```diff
@@ -0,0 +1,238 @@
+
+&gpio0 {
+	status = "okay";
+	gpio-reserved-ranges =	<6 11>, // PSRAM
```
IIRC, the `gpio-reserved-ranges` attribute is not used in ESP32 Zephyr port as this is handled by the `hal_espressif` macro(s) `SOC_GPIO_VALID_GPIO_MASK` located at the `soc_caps.h`. This macro is used by the Zephyr gpio driver.
Consider to remove the the `gpio-reserved-ranges` from the board dts.

MrMarteng replied:
Thank you. You are right.
Has been fixed

marekmatej commented:
@MrMarteng Is the M5stack and M5stack core2 that different? IIUC major differences are missing PSRAM and a few sensors on the older M5stack board. But otherwise, it seems to be very similar boards.

MrMarteng commented:
> @MrMarteng Is the M5stack and M5stack core2 that different? IIUC major differences are missing PSRAM and a few sensors on the older M5stack board. But otherwise, it seems to be very similar boards.

I am not an expert on the basic/core module, but to my assessment there are quite a view differences to core2 version. The basic has 3 only buttons, the core2 has 3 touch-buttons and a touch-screen. Core2 comes with a more advanced power-management, which was most of the work I needed to invest to get the core2 module running (#59768, #58809).

Depending if the basic module is still available on market, it might make sense to add support for this module as well. This PR only focuses on core2 module.

marekmatej commented:
> Depending if the basic module is still available on market, it might make sense to add support for this module as well. This PR only focuses on core2 module.

@MrMarteng The M5stack core is still available and is based on the same CPU as the new Core2. So I think your work is solid ground for supporting the older module if someone chooses to.

marekmatej commented on boards/xtensa/m5stack_core2/m5stack_core2.dts:
```diff
@@ -0,0 +1,237 @@
+ */
+/dts-v1/;
+
+#include <espressif/esp32/esp32_common.dtsi>
```
Please use the `esp32/esp32_wrover_e_n16r8.dtsi` instead of `esp32_common.dtsi`. There are flash and psram attributes included.

MrMarteng replied:
M5Stack-Core2 is featuring `esp32_d0wd_v3`. The implementation on the m5stack hardware is very near to `esp32_wrover_e_n16r8`, but I found it more consistent to keep sram and flash definition separately within board layer as the wrover module is not physically included. That was the reason I didn't base on the n16r8 intentionally.

marekmatej commented on boards/xtensa/m5stack_core2/m5stack_core2.dts:
```diff
@@ -0,0 +1,237 @@
+
+&flash0 {
+	status = "okay";
+	reg = <0 0x400000>;
```
Please remove the `reg` property as it has incorrect size. Correct size is defined in the `esp32_wrover_e_n16r8.dtsi`.

MrMarteng replied:
reg definition has been fixed.

MrMarteng replied:
@marekmatej Are you ok with using `esp32_d0wd_v3` as the base?

kartben commented:
> > I mean describing the pin headers similar to how it's done for say Arduino R3 headers, grove connectors, or Raspberry Pi 40-pin headers. You should be able to find examples easily by searching the code base. Thanks to this, shield definitions for M5Stack modules would conveniently be able to reference pins from the header, with Devicetree then automagically "routing" them to the corresponding pins of the ESP32. Plus you could also define aliases like m5stack_i2c or m5stack_spi.
> > Here's the description of the pin mappings for the M5Stack Core2: https://static-cdn.m5stack.com/resource/docs/products/core/core2/core2_mbus_01.webp
> > Hope this helps clarify!
> 
> Thank you. I've understood and like the idea. I will add this functionality in a separate PR.

While the "M5bus" thing can definitely come in a follow up PR, could you please add the mapping for the grove connector, since this one should be pretty trivial. See ex. https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/wio_terminal/grove_connectors.dtsi - I believe this is GPIO21 and GPIO22 that are exposed through the Grove connector.

Thank you! FWIW I am ordering a Core2 as I am typing this, and I am looking forward to putting together a video demo for my Zephyr Weekly Update as soon as this gets merged ;-)

kartben commented:
>[...] I believe this is GPIO21 and GPIO22 that are exposed through the Grove connector.

mmm actually no, that's 32 and 33.

Commit:
devicetree: Add grove connector definition for m5stack_core2

Basic pinout definition.

Signed-off-by: Martin Kiepfer <m.kiepfer@teleschirm.org>

kartben commented on boards/xtensa/m5stack_core2/m5stack_core2.dts:
```diff
@@ -0,0 +1,254 @@
+			};
+
+			lcd_bg: DCDC3 {
+				regulator-init-microvolt = <2800000>;
+				regulator-min-microvolt = <0>;
+				regulator-max-microvolt = <2800000>;
```
i would align with the min/max values from the Arduino core: min = 2500mV, max = 3300mV

https://github.com/m5stack/M5Core2/blob/master/src/AXP192.cpp#L450-L453

In any case I should say it's pretty cool to be able to tinker with the regulator shell commands to set the brightness of the LCD or disable it altogether.

MrMarteng replied:
fixed with laters commit

kartben commented on boards/xtensa/m5stack_core2/m5stack_core2.dts:
```diff
@@ -0,0 +1,254 @@
+	compatible = "espressif,esp32";
+
+	aliases {
+		pwr-led = &pwr_led;
```
it's confusing (but not blocking either I guess) that blinky wouldn't compile. Could led0 be an alias for the power LED? I couldn't seem to get it to work though but I am not really familiar with MFD (I would have expected the AXP192 GPIOs to be "just like" any other GPIO?)

MrMarteng replied:
A `gpio-leds` driver is required for the Blinky example. The `axp192_gpio1` definition within the gpio driver is required to make sure the LED is initialised during boot up (via gpis-hogs).
I added the gpio-leds definition in the latest commit. The blinky example should work now.

kartben commented on boards/xtensa/m5stack_core2/Kconfig.defconfig:
```diff
@@ -0,0 +1,42 @@
+
+config INPUT_FT5336_INTERRUPT
+	default y if INPUT
+
```
add

```
config LV_COLOR_16_SWAP
	default y if LVGL
```

otherwise LVGL sample won't look quite right out of the box

kartben replied:
I would also add, similar to other boards with a touch screen:

```
config KSCAN
	default y if DISPLAY

config INPUT
	default y if KSCAN
```

as well as an overlay for lvgl_pointer_input in samples/subsys/display/lvgl (this also requires creating an alias for ft5336 node).

when doing all of this, the LVGL demo properly switch to using a button for the Hello World (as opposed to a label), but it seems stuck. The input_dump sample seems to be working properly, so there's that.

MrMarteng replied:
fixed with latest commit

kartben replied:
cool! did you have a chance to test the LVGL sample?

kartben commented on boards/xtensa/m5stack_core2/m5stack_core2.dts:
```diff
@@ -0,0 +1,254 @@
+				regulator-init-microvolt = <2800000>;
+				regulator-min-microvolt = <0>;
+				regulator-max-microvolt = <2800000>;
+				regulator-boot-on;
```
not sure we want the LCD always on? Isn't the fact that a vin-supply is defined on the display node already enough to bind the status of the regulator to the status of the display? @gmarull

MrMarteng replied:
I am actually no longer sure if the vin-supply binding to lcd_bg regulator is correct. Is vin-supply intended to control the regulator for the control logic? ili9xx driver is not supporting brightness control. Hence I guess the vin-supply setting does not have any effect. Only a few, very specific display driver (e.v. max7219) support backlight control yet.

kartben reviewed (changes requested):
I had a chance to test this PR with actual hardware (it's a very cool board!). Please see my comments.

Commit:
drivers: axp192_gpio: Fix for latest get_direction unit test

This commit resolves an internal dependecy between GET_DIRECTION and
GET_CONFIG configuration. GET_CONFIG api is internally needed by
GET_DIRECTION api.

Signed-off-by: Martin Kiepfer <m.kiepfer@teleschirm.org>

MrMarteng commented:
Rebased to latest main.
Added additional fix for latest gpio get_direction unit test.

kartben reviewed (dismissed):
This is good to go from my perspective! Thanks @MrMarteng!
Can't wait to get I2S support on ESP32 ;-)

MrMarteng commented:
> This is good to go from my perspective! Thanks @MrMarteng!
Thanks. 

> Can't wait to get I2S support on ESP32 ;-)
Yes indeed ;-). And I would also love to see SMP working again to speed up graphical performance.

marekmatej commented:
@MrMarteng please address the `dtsi` include and fix the incorrect `m5stack-core2_defconfig` as requested by the review.

MrMarteng commented:
> @MrMarteng please address the `dtsi` include and fix the incorrect `m5stack-core2_defconfig` as requested by the review.

@marekmatej Did you see my comment on the `dtsi` include? The m5stack does not feature the wrover module. The design is just aligned on the wrover board.

marekmatej commented:
> > @MrMarteng please address the `dtsi` include and fix the incorrect `m5stack-core2_defconfig` as requested by the review.
> 
> @marekmatej Did you see my comment on the `dtsi` include? The m5stack does not feature the wrover module. The design is just aligned on the wrover board.

@MrMarteng sure, I did read it, but you probably missed my replies ;)
With the increasing number of comments, it is getting harder to read the conversation, so I asked for specific changes in the code. Please take a look at it.

kartben commented:
@marekmatej FWIW pending means that you forgot to actually submit your review :)

marekmatej commented:
> @marekmatej FWIW "pending" means that you forgot to actually submit your review so only you can see this :)

ok I see, my bad folks! but there was a few comments that were submitted and visible, but not replied :)

MrMarteng commented:
@marekmatej, @kartben Thank you both for the effort you put into this PR!
